pkg/operators/wasm/wasm.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"context"
+	_ "embed"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/tetratelabs/wazero"
+	wapi "github.com/tetratelabs/wazero/api"
+	"github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/oci"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+)
+
+const (
+	wasmObjectMediaType = "application/vnd.gadget.wasm.program.v1+binary"
+)
+
+type wasmOperator struct{}
+
+func (w *wasmOperator) Name() string {
+	return "wasm"
+}
+
+func (w *wasmOperator) Description() string {
+	return "handles wasm programs"
+}
+
+func (w *wasmOperator) InstantiateImageOperator(
+	gadgetCtx operators.GadgetContext,
+	desc ocispec.Descriptor,
+	paramValues api.ParamValues,
+) (
+	operators.ImageOperatorInstance, error,
+) {
+	return &wasmOperatorInstance{
+		desc:      desc,
+		gadgetCtx: gadgetCtx,
+		logger:    gadgetCtx.Logger(),
+	}, nil
+}
+
+type wasmOperatorInstance struct {
+	desc      ocispec.Descriptor
+	rt        wazero.Runtime
+	gadgetCtx operators.GadgetContext
+	mod       wapi.Module
+
+	logger logger.Logger
+
+	// malloc function exported by the guest
+	guestMalloc wapi.Function
+}
+
+func (i *wasmOperatorInstance) Name() string {
+	return "wasm"
+}
+
+func (i *wasmOperatorInstance) Prepare(gadgetCtx operators.GadgetContext) error {
+	err := i.init(gadgetCtx)
+	if err != nil {
+		return fmt.Errorf("initializing wasm: %w", err)
+	}
```

Comment:
Why operator initialization isn't done in the `InstantiateImageOperator` phase as for ebpf operator? 

IMO, it makes no sense to succeed in the instantiation phase but then fail in the prepare phase because we wasn't able to read the program from the image or because the wasm module doesn't export the malloc function. I see the `Prepare` more like a phase where we prepare complementary parts that the operator will use while running. For instance, for the ebpf operator, we subscribe the converters, create the tchandler, network-tracer and uprobe-tracer, but everything regarding the ebpf operator itself was already initialized before. 

Additionally, It'll prevent us from carrying the oras targer and image descriptor in `wasmOperatorInstance`, but just the WASM module already instantiated.

Reply:
It's not totally clear to me what are different phases for. Let's ask @flyth about this one.

Reply:
I somehow thought I had already answered to this, sorry.

[This](https://gist.github.com/flyth/d98befcdd7a1c6558af9af88012ee89a#operators) has a short overview of the lifecycle hooks and what should be done when. It's crucial that this operator forwards the `init` in the instantiate phase, as that is the only phase that will be called for `GetGadgetInfo()`. It is the phase in which all new data sources and alterations to existing ones need to be done in order for all operators to pick them up. The `Prepare()` phase is an optional phase that you can use to subscribe to datasources - it should be forwarded to Wasm as well.

Reply:
> It's crucial that this operator forwards the init in the instantiate phase, as that is the only phase that will be called for GetGadgetInfo()

Are you totally sure about this? I made a simple experiment and it seems to me Prepare() is also called on GetGadgetInfo()

https://github.com/inspektor-gadget/inspektor-gadget/blob/556fc13b14f5b899766f2013f26d85a548d27e9b/pkg/gadget-service/service-oci.go#L61

https://github.com/inspektor-gadget/inspektor-gadget/blob/5813c0087f3dda1ad7dee75f26108bc8013bf355/pkg/runtime/local/oci.go#L26

https://github.com/inspektor-gadget/inspektor-gadget/blob/6e250d5258598d56f206ddb8d99e911666f81e19/pkg/gadget-context/run.go#L148

https://github.com/inspektor-gadget/inspektor-gadget/blob/6e250d5258598d56f206ddb8d99e911666f81e19/pkg/gadget-context/run.go#L53

https://github.com/inspektor-gadget/inspektor-gadget/blob/e5fed8d72b088c56282e09dc0acd7ec3cfdfc663/pkg/operators/oci-handler/oci.go#L150

https://github.com/inspektor-gadget/inspektor-gadget/blob/e5fed8d72b088c56282e09dc0acd7ec3cfdfc663/pkg/operators/oci-handler/oci.go#L265

Is it a bug on the oci-handler operator?

In any case, I'll move this to InstantiateImageOperator() as it makes more sense after your explanation.